### PR TITLE
Update RayTrace Overide Patch to 1.14

### PR DESCRIPTION
**Pull request by ShetiPhian**
1.14 Update for Patch: https://github.com/MinecraftForge/MinecraftForge/pull/5354

Short Function Summery:
Enables sub segment block outlines, and assignment of the 'subHit' variable

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java [hunk 1]
```diff
@@ -988,11 +990,11 @@ else if ((entity instanceof WitherEntity) ||
      * @param pos Block position in world
      * @param start The start vector
      * @param end The end vector
-     * @param original The original result from {@link Block#collisionRayTrace(IBlockState, World, BlockPos, Vec3d, Vec3d)}
+     * @param original The original result from {@link IBlockReader#func_217296_a(Vec3d, Vec3d, BlockPos, VoxelShape, BlockState)}
      * @return A result that suits your block
      */
     @Nullable
-    default RayTraceResult getRayTraceResult(BlockState state, World world, BlockPos pos, Vec3d start, Vec3d end, RayTraceResult original)
+    default BlockRayTraceResult getRayTraceResult(BlockState state, IBlockReader world, BlockPos pos, Vec3d start, Vec3d end, BlockRayTraceResult original)
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:1008]: We do not allow breaking changes, but I fear this one is neccessary, as we cannot safly upcast the IBlockReader and RayTraceResult. Also, mods that simply override this function will not break, and mods should not call this function manually...

Comment (diesieben07) [reply src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:1008]: > mods that simply override this function will not break

they will break if they call super.
So the method should ideally be kept and deprecated.

Comment (ichttt) [reply src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:1008]: yeah right, forgot it was defaulted...

**Comment by Elec332** [comment]
Is there anything preventing this from being merged?

I would love to be able to assign subHit again :)
Multipart-like blocks need this to work properly.

**Comment by ShetiPhian** [comment]
@Elec332 
Forget this pull, and ignore subHit completely.

This code will take an array of VoxelShapes and return the index of the one with the shortest traced distance. 
```
public static int getSubShapeHit(Entity entity, BlockPos pos, VoxelShape... shapes)
{
	double distance = 5;
	Vec3d vecStart = new Vec3d(entity.prevPosX, entity.prevPosY + entity.getEyeHeight(), entity.prevPosZ);
	Vec3d vecLook = entity.getLook(0);
	Vec3d vecEnd = vecStart.add(vecLook.x * distance, vecLook.y * distance, vecLook.z * distance);

	int shapeIndex = -1;
	double hitDistance = 100;
	for (byte index = 0; index < shapes.length; index++) {
		BlockRayTraceResult shapeHit = shapes[index].rayTrace(vecStart, vecEnd, pos);
		if (shapeHit == null) {
			continue;
		}
		double shapeDistance = shapeHit.getHitVec().squareDistanceTo(vecStart);
		if (shapeDistance < hitDistance) {
			hitDistance = shapeDistance;
			shapeIndex = index;
		}
	}
	return shapeIndex;
}
```

Use DrawBlockHighlightEvent to render the outline.
copy the code from WorldRenderer.drawSelectionBox and change it so you can pass the VoxelShape at the index provided by getSubShapeHit


Call getSubShapeHit anywhere you need to handle interactions.

**Pull request closed without merging**